File: clients/go/dlq.go
package wavehouse

import (
	"context"
	"fmt"
	"net/url"
)

// DLQNamespace provides admin-only dead-letter-queue statistics.
//
// The server registers /v1/ops/dlq/stats only when the DLQ is enabled, so on a
// deployment with dlq.enabled: false these calls return an [*Error] with
// Status 404 — "the DLQ is switched off", not "the DLQ is empty". Check
// Status before reading a zero DLQStats as a healthy result.
type DLQNamespace struct {
	ctx          httpContext
	createStream func(table string, opts *StreamOptions) *StreamController
}

// List returns DLQ statistics (message counts per table). Admin-only.
func (d *DLQNamespace) List(ctx context.Context) (*DLQStats, error) {
	return d.stats(ctx, nil)
}

// Table returns DLQ stats filtered by table name. Admin-only.
func (d *DLQNamespace) Table(ctx context.Context, name string) (*DLQStats, error) {
	return d.stats(ctx, url.Values{"table": {name}})
}

func (d *DLQNamespace) stats(ctx context.Context, params url.Values) (*DLQStats, error) {
	var stats DLQStats
	if err := doRequest(ctx, d.ctx, requestOptions{
		method: "GET",
		path:   "/v1/ops/dlq/stats",
		params: params,
	}, &stats); err != nil {
		return nil, fmt.Errorf("get dlq stats: %w", err)
	}
	return &stats, nil
}

// Stream subscribes to live DLQ events. Not yet functional server-side (#197).
func (d *DLQNamespace) Stream(opts *StreamOptions) *StreamController {
	return d.createStream("dlq", opts)
}

File: clients/ts/src/dlq.ts
import { err, ok } from "./errors.js";
import { request, tenantParam } from "./http.js";
import type { StreamController } from "./stream/controller.js";
import type { DLQStats, HttpContext, OpsRequestOptions, Result, StreamOptions } from "./types.js";

type CreateStreamFn = (table: string, opts?: StreamOptions) => StreamController;

/** Namespace for Dead Letter Queue operations. */
export class DLQNamespace {
  private readonly _ctx: HttpContext;
  private readonly _createStream: CreateStreamFn;

  constructor(ctx: HttpContext, createStream: CreateStreamFn) {
    this._ctx = ctx;
    this._createStream = createStream;
  }

  /**
   * Get DLQ statistics (message counts per table) — of `opts.tenant`, the
   * default tenant without it. A tenant with no dead-letter queue is a `404`.
   */
  async list(opts?: OpsRequestOptions): Promise<Result<DLQStats>> {
    const { data, error } = await request<DLQStats>(this._ctx, {
      method: "GET",
      path: "/v1/ops/dlq/stats",
      params: tenantParam(opts),
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }

  /** Get DLQ stats filtered by table name — of `opts.tenant`, the default tenant without it. */
  async table(name: string, opts?: OpsRequestOptions): Promise<Result<DLQStats>> {
    const { data, error } = await request<DLQStats>(this._ctx, {
      method: "GET",
      path: "/v1/ops/dlq/stats",
      params: { table: name, ...tenantParam(opts) },
      signal: opts?.signal,
    });
    if (error) return err(error);
    return ok(data!);
  }

  /** Subscribe to live DLQ events. */
  stream(opts?: StreamOptions): StreamController {
    return this._createStream("dlq", opts);
  }
}
